File: producer-consumer-rabbitmq/producer-js/producer.js
const amqp = require("amqplib");

const QUEUES = ["tareas", "deportes", "educacion"];
const RABBITMQ_HOST = "rabbitmq";
const RABBITMQ_USER = "user";
const RABBITMQ_PASS = "password";
const RETRY_DELAY = 5000; // 5 segundos

async function connectRabbitMQ() {
  while (true) {
    try {
      console.log("[*] Intentando conectar a RabbitMQ...");
      const connection = await amqp.connect({
        protocol: "amqp",
        hostname: RABBITMQ_HOST,
        port: 5672,
        username: RABBITMQ_USER,
        password: RABBITMQ_PASS,
      });
      console.log("[✔] Conectado a RabbitMQ desde Producer JS");
      return connection;
    } catch (err) {
      console.error("[!] RabbitMQ no disponible. Reintentando en 5s...");
      await new Promise((resolve) => setTimeout(resolve, RETRY_DELAY));
    }
  }
}

async function publishMessages() {
  try {
    const connection = await connectRabbitMQ();
    const channel = await connection.createChannel();

    // Asegurar todas las colas
    for (const queue of QUEUES) {
      await channel.assertQueue(queue, { durable: false });
    }

    let counter = 0;
    setInterval(() => {
      for (const queue of QUEUES) {
        const msg = `Mensaje ${counter} desde Producer JS a ${queue}`;
        channel.sendToQueue(queue, Buffer.from(msg));
        console.log("[x] Enviado:", msg);
      }
      counter++;
    }, 3000);

    // Manejo si RabbitMQ se cae
    connection.on("close", () => {
      console.error("[!] Conexión a RabbitMQ perdida. Reconectando...");
      setTimeout(publishMessages, RETRY_DELAY);
    });
    connection.on("error", (err) => {
      console.error("[!] Error en conexión RabbitMQ:", err.message);
    });

  } catch (err) {
    console.error("Error crítico en Producer JS:", err);
    setTimeout(publishMessages, RETRY_DELAY);
  }
}

publishMessages();

File: producer-consumer-rabbitmq/producer-java/src/main/java/com/example/Producer.java
package com.example;

import com.rabbitmq.client.Channel;
import com.rabbitmq.client.Connection;
import com.rabbitmq.client.ConnectionFactory;

public class Producer {
    private static final String[] QUEUES = {"tareas", "deportes", "educacion"};
    private static final String HOST = "rabbitmq";
    private static final String USER = "user";
    private static final String PASS = "password";
    private static final int RETRY_DELAY = 5000; // ms

    public static void main(String[] argv) {
        ConnectionFactory factory = new ConnectionFactory();
        factory.setHost(HOST);
        factory.setUsername(USER);
        factory.setPassword(PASS);

        while (true) {
            try (Connection connection = factory.newConnection();
                 Channel channel = connection.createChannel()) {

                System.out.println("[✔] Conectado a RabbitMQ desde Producer Java");

                // Asegurar colas
                for (String queue : QUEUES) {
                    channel.queueDeclare(queue, false, false, false, null);
                }

                int counter = 0;
                while (true) {
                    for (String queue : QUEUES) {
                        String message = "Mensaje " + counter + " desde Producer Java a " + queue;
                        channel.basicPublish("", queue, null, message.getBytes());
                        System.out.println("[x] Enviado: " + message);
                    }
                    counter++;
                    Thread.sleep(3000);
                }

            } catch (Exception e) {
                System.err.println("[!] Error conectando o enviando: " + e.getMessage());
                try {
                    Thread.sleep(RETRY_DELAY);
                } catch (InterruptedException ignored) {
                }
                System.out.println("[*] Reintentando conexión...");
            }
        }
    }
}
